src/track/cue.cpp

```diff
@@ -44,15 +45,52 @@ Cue::Cue(int id, TrackId trackId, Cue::Type type, double position, double length
           m_color(color) {
     DEBUG_ASSERT(!m_label.isNull());
     if (length) {
-        if (position != Cue::kNoPosition) {
+        if (position != mixxx::kNoPosition) {
             m_sampleEndPosition = position + length;
         } else {
             m_sampleEndPosition = length;
         }
     } else {
-        m_sampleEndPosition = Cue::kNoPosition;
+        m_sampleEndPosition = mixxx::kNoPosition;
     }
 }
+
+Cue::Cue(TrackId trackId, int sampleRateHz, const mixxx::CueInfo& cueInfo)
+        : m_bDirty(false),
+          m_iId(-1),
+          m_trackId(trackId),
+          m_type(cueInfo.getType()),
+          m_sampleStartPosition(mixxx::kNoPosition),
+          m_sampleEndPosition(mixxx::kNoPosition),
+          m_iHotCue(cueInfo.getHotCueNumber()),
+          m_label(cueInfo.getLabel()),
+          m_color(Color::kPredefinedColorsSet.noColor) {
+    DEBUG_ASSERT(!m_label.isNull());
+
+    const double sampleRateKhz = sampleRateHz / 1000.0;
+    const double millisecsToSamplesFactor = sampleRateKhz * mixxx::kEngineChannelCount;
+    DEBUG_ASSERT(millisecsToSamplesFactor > 0);
+
+    if (cueInfo.getStartPositionMillis() != mixxx::kNoPosition) {
+        m_sampleStartPosition = cueInfo.getStartPositionMillis() * millisecsToSamplesFactor;
+    }
+
+    if (cueInfo.getEndPositionMillis() != mixxx::kNoPosition) {
+        m_sampleEndPosition = cueInfo.getEndPositionMillis() * millisecsToSamplesFactor;
+    }
+
+    mixxx::RgbColor::optional_t color = cueInfo.getColor();
+    if (color) {
+        // All tracks have the same color
```

Comment:
This generic mapping of a color to a predefined color does not belong here where it couldn't be reused. Extract functions and add them to PredefinedColorsSet:

```
PredefinedColorPointer mapToPredefinedColor(mixxx::RgbColor color);
PredefinedColorPointer mapToPredefinedColor(mixxx::RgbColor::optional_t color); 
```

If no matching color is found then `PredefinedColorPointer()` should be returned. The second variant will map `nullopt` to `noColor`.

Reply:
With #2520 this is moot anyway, but move that mapping function to `PredefinedColorsSet`.